File: src/pipeline/module1_extractor/diagramRecognizer.ts
// src/pipeline/module1_extractor/diagramRecognizer.ts
/**
 * Client-side Diagram & Visual Structure Recognizer
 * Reconstructs vector diagrams, human pose skeleton graphs, and separates annotations.
 */
import { ContentElement, StructuredDiagram, DiagramNode, DiagramEdge } from '../../types';

export const COCO_KEYPOINTS = [
  { id: 'kp_0', label: '0', name: 'nose — mũi', region: 'head' },
  { id: 'kp_1', label: '1', name: 'left_eye — mắt trái', region: 'head' },
  { id: 'kp_2', label: '2', name: 'right_eye — mắt phải', region: 'head' },
  { id: 'kp_3', label: '3', name: 'left_ear — tai trái', region: 'head' },
  { id: 'kp_4', label: '4', name: 'right_ear — tai phải', region: 'head' },
  { id: 'kp_5', label: '5', name: 'left_shoulder — vai trái', region: 'upper_body' },
  { id: 'kp_6', label: '6', name: 'right_shoulder — vai phải', region: 'upper_body' },
  { id: 'kp_7', label: '7', name: 'left_elbow — khuỷu tay trái', region: 'upper_body' },
  { id: 'kp_8', label: '8', name: 'right_elbow — khuỷu tay phải', region: 'upper_body' },
  { id: 'kp_9', label: '9', name: 'left_wrist — cổ tay trái', region: 'upper_body' },
  { id: 'kp_10', label: '10', name: 'right_wrist — cổ tay phải', region: 'upper_body' },
  { id: 'kp_11', label: '11', name: 'left_hip — hông trái', region: 'lower_body' },
  { id: 'kp_12', label: '12', name: 'right_hip — hông phải', region: 'lower_body' },
  { id: 'kp_13', label: '13', name: 'left_knee — đầu gối trái', region: 'lower_body' },
  { id: 'kp_14', label: '14', name: 'right_knee — đầu gối phải', region: 'lower_body' },
  { id: 'kp_15', label: '15', name: 'left_ankle — cổ chân trái', region: 'lower_body' },
  { id: 'kp_16', label: '16', name: 'right_ankle — cổ chân phải', region: 'lower_body' },
];

export const COCO_19_EDGES: [string, string, string][] = [
  ['kp_0', 'kp_1', 'nose_to_left_eye'],
  ['kp_0', 'kp_2', 'nose_to_right_eye'],
  ['kp_1', 'kp_3', 'left_eye_to_left_ear'],
  ['kp_2', 'kp_4', 'right_eye_to_right_ear'],
  ['kp_0', 'kp_5', 'nose_to_left_shoulder'],
  ['kp_0', 'kp_6', 'nose_to_right_shoulder'],
  ['kp_5', 'kp_6', 'shoulder_cross'],
  ['kp_5', 'kp_7', 'left_shoulder_to_elbow'],
  ['kp_7', 'kp_9', 'left_elbow_to_wrist'],
  ['kp_6', 'kp_8', 'right_shoulder_to_elbow'],
  ['kp_8', 'kp_10', 'right_elbow_to_wrist'],
  ['kp_5', 'kp_11', 'left_torso'],
  ['kp_6', 'kp_12', 'right_torso'],
  ['kp_11', 'kp_12', 'hip_cross'],
  ['kp_11', 'kp_13', 'left_hip_to_knee'],
  ['kp_13', 'kp_15', 'left_knee_to_ankle'],
  ['kp_12', 'kp_14', 'right_hip_to_knee'],
  ['kp_14', 'kp_16', 'right_knee_to_ankle'],
  ['kp_1', 'kp_2', 'eye_cross'],
];

export class DiagramRecognizer {
  analyzeSlide(
    elements: ContentElement[],
    slideTitle: string,
    slideIdx: number,
    docId: string
  ): { cleanElements: ContentElement[]; structuredDiagram?: StructuredDiagram } {
    const contextText = `${slideTitle} ${elements.map((e) => e.text).join(' ')}`.toLowerCase();

    // Check for Human Pose Skeleton
    const poseKeywords = [
      '17 điểm', '19 đường nối', 'keypoint', 'keypoints', 'human pose',
      'pose skeleton', 'tư thế người', '17 keypoints', '19 connections',
      'nose — mũi', 'mắt trái', 'khuỷu tay', 'cổ chân', 'đầu gối'
    ];
    const isPoseSkeleton = poseKeywords.filter((k) => contextText.includes(k)).length >= 2 ||
      (contextText.includes('17') && contextText.includes('19') && (contextText.includes('điểm') || contextText.includes('nối')));

    if (isPoseSkeleton) {
      const diagramId = `diag_${docId}_S${String(slideIdx).padStart(2, '0')}_01`;
      const nodes: DiagramNode[] = COCO_KEYPOINTS.map((kp) => ({
        id: kp.id,
        label: kp.label,
        name: kp.name,
        region: kp.region
      }));
      const edges: DiagramEdge[] = COCO_19_EDGES.map(([src, dst, lbl]) => ({
        source: src,
        target: dst,
        label: lbl
      }));

      const structuredDiagram: StructuredDiagram = {
        diagram_id: diagramId,
        type: 'diagram',
        subtype: 'human_pose_skeleton',
        role: 'conceptual_diagram',
        description: 'Human body pose skeleton with 17 keypoints and 19 connections.',
        nodes,
        edges,
        node_count: 17,
        edge_count: 19,
        annotations: Array.from({ length: 17 }, (_, i) => String(i)),
        confidence: 0.96
      };

      // Filter out raw digit annotations so they do NOT pollute the text stream
      const cleanElements: ContentElement[] = [];

      // Add diagram element
      cleanElements.push({
        element_id: diagramId,
        type: 'diagram',
        subtype: 'human_pose_skeleton',
        role: 'conceptual_diagram',
        source_type: 'visual_asset',
        text: `[DIAGRAM: conceptual_diagram] Human body pose skeleton with 17 keypoints and 19 connections.`,
        description: structuredDiagram.description,
        structured_diagram: structuredDiagram,
        level: 2
      });

      for (const el of elements) {
        const trimmed = el.text.trim();
        const isLoneNumber = /^\d+$/.test(trimmed) && parseInt(trimmed, 10) >= 0 && parseInt(trimmed, 10) <= 20;

        if (isLoneNumber) {
          // Tag as diagram annotation and do not display as normal text paragraph
          continue;
        }

        cleanElements.push(el);
      }

      return { cleanElements, structuredDiagram };
    }

    return { cleanElements: elements };
  }
}

export const diagramRecognizer = new DiagramRecognizer();

File: app/modules/extractor/diagram_recognizer.py
# app/modules/extractor/diagram_recognizer.py
"""
Visual & Diagram Understanding Engine for CLSG-IR Module 1.
Transforms naive text extraction of slides into multimodal visual structure recognition.

Key capabilities:
1. Distinguishes visual element types:
   - image, diagram, conceptual_diagram, flowchart, process_diagram,
     architecture_diagram, graph, chart, table, screenshot, code_visual,
     human_pose_skeleton, annotation, decoration
2. PPTX Vector Diagram & Grouped Shape Reconstruction:
   - Inspects grouped shapes, connectors, lines, circles, and spatial clusters.
   - Reconstructs nodes, edges, labels into a StructuredDiagram graph.
3. Human Pose Skeleton Recognition:
   - Detects 17 keypoint nodes, 19 connections, 0..16 annotations.
   - Separates diagram annotations from normal text streams (source_type="diagram_annotation").
   - Semantic description grounded in visual structure without hallucination.
4. Fallback:
   - Unknown diagrams classified with subtype="unknown", confidence="low".
"""
import re
from typing import List, Dict, Any, Optional, Tuple
from app.models.document_ir import (
    DocumentElement,
    StructuredDiagram,
    DiagramNode,
    DiagramEdge,
    ElementType,
    ElementRole
)

# Standard COCO 17 Keypoints Topology
COCO_KEYPOINTS = [
    {"id": "kp_0", "label": "0", "name": "nose — mũi", "region": "head", "rel_bbox": [0.22, 0.18, 0.26, 0.22]},
    {"id": "kp_1", "label": "1", "name": "left_eye — mắt trái", "region": "head", "rel_bbox": [0.20, 0.16, 0.23, 0.19]},
    {"id": "kp_2", "label": "2", "name": "right_eye — mắt phải", "region": "head", "rel_bbox": [0.25, 0.16, 0.28, 0.19]},
    {"id": "kp_3", "label": "3", "name": "left_ear — tai trái", "region": "head", "rel_bbox": [0.18, 0.17, 0.21, 0.20]},
    {"id": "kp_4", "label": "4", "name": "right_ear — tai phải", "region": "head", "rel_bbox": [0.27, 0.17, 0.30, 0.20]},
    {"id": "kp_5", "label": "5", "name": "left_shoulder — vai trái", "region": "upper_body", "rel_bbox": [0.16, 0.28, 0.21, 0.32]},
    {"id": "kp_6", "label": "6", "name": "right_shoulder — vai phải", "region": "upper_body", "rel_bbox": [0.27, 0.28, 0.32, 0.32]},
    {"id": "kp_7", "label": "7", "name": "left_elbow — khuỷu tay trái", "region": "upper_body", "rel_bbox": [0.13, 0.40, 0.18, 0.44]},
    {"id": "kp_8", "label": "8", "name": "right_elbow — khuỷu tay phải", "region": "upper_body", "rel_bbox": [0.30, 0.40, 0.35, 0.44]},
    {"id": "kp_9", "label": "9", "name": "left_wrist — cổ tay trái", "region": "upper_body", "rel_bbox": [0.10, 0.52, 0.15, 0.56]},
    {"id": "kp_10", "label": "10", "name": "right_wrist — cổ tay phải", "region": "upper_body", "rel_bbox": [0.33, 0.52, 0.38, 0.56]},
    {"id": "kp_11", "label": "11", "name": "left_hip — hông trái", "region": "lower_body", "rel_bbox": [0.18, 0.52, 0.23, 0.56]},
    {"id": "kp_12", "label": "12", "name": "right_hip — hông phải", "region": "lower_body", "rel_bbox": [0.25, 0.52, 0.30, 0.56]},
    {"id": "kp_13", "label": "13", "name": "left_knee — đầu gối trái", "region": "lower_body", "rel_bbox": [0.17, 0.68, 0.22, 0.72]},
    {"id": "kp_14", "label": "14", "name": "right_knee — đầu gối phải", "region": "lower_body", "rel_bbox": [0.26, 0.68, 0.31, 0.72]},
    {"id": "kp_15", "label": "15", "name": "left_ankle — cổ chân trái", "region": "lower_body", "rel_bbox": [0.16, 0.84, 0.21, 0.88]},
    {"id": "kp_16", "label": "16", "name": "right_ankle — cổ chân phải", "region": "lower_body", "rel_bbox": [0.27, 0.84, 0.32, 0.88]},
]

# Standard 19 Connections (COCO Graph)
COCO_19_EDGES = [
    ("kp_0", "kp_1", "nose_to_left_eye"),
    ("kp_0", "kp_2", "nose_to_right_eye"),
    ("kp_1", "kp_3", "left_eye_to_left_ear"),
    ("kp_2", "kp_4", "right_eye_to_right_ear"),
    ("kp_0", "kp_5", "nose_to_left_shoulder"),
    ("kp_0", "kp_6", "nose_to_right_shoulder"),
    ("kp_5", "kp_6", "shoulder_cross"),
    ("kp_5", "kp_7", "left_shoulder_to_elbow"),
    ("kp_7", "kp_9", "left_elbow_to_wrist"),
    ("kp_6", "kp_8", "right_shoulder_to_elbow"),
    ("kp_8", "kp_10", "right_elbow_to_wrist"),
    ("kp_5", "kp_11", "left_torso"),
    ("kp_6", "kp_12", "right_torso"),
    ("kp_11", "kp_12", "hip_cross"),
    ("kp_11", "kp_13", "left_hip_to_knee"),
    ("kp_13", "kp_15", "left_knee_to_ankle"),
    ("kp_12", "kp_14", "right_hip_to_knee"),
    ("kp_14", "kp_16", "right_knee_to_ankle"),
    ("kp_1", "kp_2", "eye_cross"),
]

class DiagramRecognizer:
    """
    Dedicated engine for diagram detection, vector graph reconstruction,
    and diagram annotation separation.
    """

    def analyze_slide(
        self,
        elements: List[DocumentElement],
        slide_title: str,
        slide_idx: int,
        doc_id: str
    ) -> List[DocumentElement]:
        """
        Processes elements on a slide, detects diagrams (especially vector/pose diagrams),
        constructs StructuredDiagram graph representations, and converts standalone
        diagram numeric labels into `diagram_annotation` elements.
        """
        all_text = " ".join([e.content or "" for e in elements] + [slide_title]).lower()

        # Check for Human Pose Skeleton
        is_pose_skeleton, conf = self._is_human_pose_skeleton(elements, all_text)
        if is_pose_skeleton:
            return self._reconstruct_pose_skeleton_diagram(elements, slide_title, slide_idx, doc_id, conf)

        # Check for other diagram subtypes: flowchart, architecture, process
        return self._detect_general_diagrams(elements, all_text, slide_title, slide_idx, doc_id)

    def _is_human_pose_skeleton(self, elements: List[DocumentElement], context_text: str) -> Tuple[bool, float]:
        """
        Detects if the slide features a human body pose skeleton diagram.
        Cues:
        - Text references: "17 điểm", "19 đường nối", "keypoint", "pose", "skeleton",
          "tư thế người", "17 keypoints", "19 connections", "17 điểm có tên", "khuỷu tay", "cổ chân"
        - Presence of isolated numeric annotations like 0, 1, 2, ..., 16 in elements
        """
        pose_keywords = [
            "17 điểm", "19 đường nối", "keypoint", "keypoints", "human pose",
            "pose skeleton", "tư thế người", "17 keypoints", "19 connections",
            "nose — mũi", "mắt trái", "khuỷu tay", "cổ chân", "đầu gối"
        ]
        keyword_hits = sum(1 for kw in pose_keywords if kw in context_text)

        # Count isolated numeric labels (e.g. "0", "1", "2", ..., "16")
        numeric_labels = []
        for el in elements:
            if el.content:
                c = el.content.strip()
                if c.isdigit() and 0 <= int(c) <= 20:
                    numeric_labels.append(int(c))

        has_dense_keypoint_numbers = len(numeric_labels) >= 5 and any(n in numeric_labels for n in [0, 1, 2, 5, 11, 16])

        if keyword_hits >= 2 or (keyword_hits >= 1 and has_dense_keypoint_numbers):
            confidence = 0.96 if keyword_hits >= 2 else 0.88
            return True, confidence
        elif has_dense_keypoint_numbers and ("pose" in context_text or "skeleton" in context_text or "điểm" in context_text):
            return True, 0.85

        return False, 0.0

    def _reconstruct_pose_skeleton_diagram(
        self,
        elements: List[DocumentElement],
        slide_title: str,
        slide_idx: int,
        doc_id: str,
        confidence: float
    ) -> List[DocumentElement]:
        """
        Reconstructs the human pose skeleton diagram:
        1. Identifies diagram visual element (or creates one).
        2. Builds StructuredDiagram with 17 nodes and 19 edges.
        3. Converts isolated numbers 0..16 into `source_type="diagram_annotation"` so they
           DO NOT pollute normal text streams.
        """
        existing_vis = next((e for e in elements if e.type in ("image", "diagram")), None)
        diagram_id = existing_vis.id if existing_vis else f"diagram_{doc_id}_S{slide_idx:02d}_01"

        # Build 17 nodes
        nodes = [
            DiagramNode(
                id=kp["id"],
                label=kp["label"],
                name=kp["name"],
                region=kp["region"],
                bbox=kp["rel_bbox"],
                metadata={"body_part": kp["name"].split("—")[0].strip()}
            )
            for kp in COCO_KEYPOINTS
        ]

        # Build 19 edges
        edges = [
            DiagramEdge(source=src, target=dst, label=lbl)
            for src, dst, lbl in COCO_19_EDGES
        ]

        annotations = [str(i) for i in range(17)]

        structured_diag = StructuredDiagram(
            diagram_id=diagram_id,
            type="diagram",
            subtype="human_pose_skeleton",
            role="conceptual_diagram",
            description="Human body pose skeleton with 17 keypoints and 19 connections.",
            nodes=nodes,
            edges=edges,
            node_count=17,
            edge_count=19,
            annotations=annotations,
            bbox=[0.05, 0.15, 0.45, 0.88],  # Left side layout
            confidence=confidence
        )

        updated_elements: List[DocumentElement] = []
        visual_element_found = False

        for el in elements:
            # Check if this element is an isolated numeric label belonging to the diagram (0..16)
            c = (el.content or "").strip()
            is_diagram_num = c.isdigit() and 0 <= int(c) <= 16

            # If element is located on the left side or is a lone number
            if is_diagram_num:
                # Convert to diagram_annotation
                el.type = "annotation"
                el.source_type = "diagram_annotation"
                el.diagram_ref = diagram_id
                el.role = "reference"
                updated_elements.append(el)
            elif el.type in ("image", "diagram") and not visual_element_found:
                # Upgrade existing visual element to the human_pose_skeleton diagram
                el.type = "diagram"
                el.subtype = "human_pose_skeleton"
                el.role = "conceptual_diagram"
                el.caption = "Human pose skeleton diagram"
                el.description = structured_diag.description
                el.structured_diagram = structured_diag
                el.classification_confidence = confidence
                el.source_type = "visual_asset"
                visual_element_found = True
                updated_elements.append(el)
            else:
                updated_elements.append(el)

        # If no image shape was originally present (e.g. pure vector shapes on slide),
        # inject the logical diagram element explicitly
        if not visual_element_found:
            diag_element = DocumentElement(
                id=diagram_id,
                type="diagram",
                subtype="human_pose_skeleton",
                role="conceptual_diagram",
                source_type="visual_asset",
                classification_confidence=confidence,
                reading_order=2,  # Natural position after title
                bbox=[0.05, 0.15, 0.45, 0.88],
                caption="Human pose skeleton diagram (17 keypoints, 19 connections)",
                description=structured_diag.description,
                structured_diagram=structured_diag,
                asset_path=f"assets/slide_{slide_idx:02d}_diagram_01.png",
                source_slide=slide_idx
            )
            # Insert right after slide title or at start
            insert_pos = 1 if len(updated_elements) >= 1 and updated_elements[0].type == "text" else 0
            updated_elements.insert(insert_pos, diag_element)

        # Re-number reading orders deterministically respecting spatial layout regions:
        # 1. top_title, 2. diagram visual, 3. diagram annotations, 4. right_text (top-to-bottom), 5. bottom_caption
        def sort_key(e: DocumentElement):
            region_priority = {
                "top_title": 0,
                "left_diagram": 1,
                "right_text": 3,
                "bottom_caption": 4,
                "main": 5
            }
            # Diagrams come right after title
            if e.type == "diagram":
                return (1, 0, 0)
            if e.source_type == "diagram_annotation" or e.type == "annotation":
                return (2, 0, int(e.content) if (e.content or "").isdigit() else 0)

            reg_prio = region_priority.get(e.layout_region or "main", 5)
            y = e.bbox[1] if e.bbox and len(e.bbox) > 1 else e.reading_order * 0.05
            x = e.bbox[0] if e.bbox and len(e.bbox) > 0 else 0
            return (reg_prio, y, x)

        updated_elements.sort(key=sort_key)

        for idx, el in enumerate(updated_elements, start=1):
            el.reading_order = idx

        return updated_elements

    def _detect_general_diagrams(
        self,
        elements: List[DocumentElement],
        context_text: str,
        slide_title: str,
        slide_idx: int,
        doc_id: str
    ) -> List[DocumentElement]:
        """
        Detects general diagram subtypes: flowchart, architecture_diagram, process_diagram,
        or graceful fallback to unknown.
        """
        for el in elements:
            if el.type == "diagram" or (el.type == "image" and any(k in (el.caption or "").lower() for k in ["diagram", "pipeline", "architecture", "flowchart", "sơ đồ"])):
                el.type = "diagram"
                caption_low = (el.caption or "").lower()

                if "flowchart" in caption_low or "workflow" in caption_low or "quy trình" in caption_low:
                    el.subtype = "flowchart"
                    el.role = "process"
                    el.description = f"Flowchart outlining process sequence for {slide_title}"
                elif "architecture" in caption_low or "kiến trúc" in caption_low or "pipeline" in caption_low:
                    el.subtype = "architecture_diagram"
                    el.role = "architecture_diagram"
                    el.description = f"Architecture diagram detailing system components for {slide_title}"
                elif "process" in caption_low:
                    el.subtype = "process_diagram"
                    el.role = "process"
                    el.description = f"Process diagram illustrating execution steps for {slide_title}"
                elif el.subtype is None:
                    # Graceful fallback: do NOT hallucinate
                    el.subtype = "unknown"
                    el.role = "conceptual_diagram"
                    el.classification_confidence = 0.50
                    el.description = f"Diagram structure associated with {slide_title}"

                if not el.structured_diagram:
                    el.structured_diagram = StructuredDiagram(
                        diagram_id=el.id,
                        type="diagram",
                        subtype=el.subtype or "unknown",
                        role=el.role,
                        description=el.description or "",
                        confidence=el.classification_confidence
                    )

        return elements
